### ISSUE TITLE
Allow Leo to use black :-)

### ISSUE THREAD

**Issue by edreamleo** [psf/black#3397]
I would love to have [Leo](http://leoeditor.com/) use Black as Leo's one-and-only code formatter. That's not possible at present.

**The problem**

Leo uses **sentinels**, comments starting with `#@` (no space after the pound sign), to denote outline structure. Sentinels appear in the **external files** that Leo produces.

This convention goes back more than two decades. Recent versions of Leo can *read* external files with black-formatted sentinels, but *writing* black-formatted sentinels would make it impossible for older versions of Leo to read external files.

In short, Leo can not allow Black to change sentinel comments.

**A new option?**

I propose an option that tells Black not to format Leo sentinels :-) Perhaps a regex could describe comments that Black should leave alone.

I won't go into details unless the Black community is interested.

**Note**: Leo's users never see sentinel comments. This proposal would have *no* effect on Black's code style.

**Comment by JelleZijlstra** [comment]
This would be implementable by changing https://github.com/psf/black/blob/main/src/black/comments.py#L32, which controls comments that don't get spaces inserted.

If we stop adding spaces to '#@`, that could affect non-Leo users. Python lines can start with `@` (decorators), so if a user comments out a decorator by just adding `#`, currently Black will clean that up by adding a space.

A new option would fix this, but we're hesitant to introduce new options that affect formatting.

**Comment by edreamleo** [comment]
@JelleZijlstra Thanks for your comments.

Oh joy. Leo (and related scripts) need only monkey-patch the comments.COMMENT_EXCEPTIONS dict. This plan should work "forever" provided that this small part of black's code remains unchanged.

Would it make sense to add a comment stating that Leo expects the COMMENT_EXCEPTIONS dict to exist?

I'll close this issue for now, and will reopen only if unexpected difficulties arise. Many thanks for your help.

**Comment by hauntsaninja** [comment]
Note that monkey patching black may not be simple in the event that you use the mypyc-compiled wheels from PyPI. 

You may be interested in stealing some of the code from https://github.com/grantjenks/blue that messes with import loading to monkey patch. You can also always install with --no-binary.

**Comment by edreamleo** [comment]
@hauntsaninja Thanks for your helpful comment.

**Comment by edreamleo** [comment]
@JelleZijlstra @hauntsaninja

Monkey-patching COMMENT_EXCEPTIONS dict will not suffice for Leo.

Happily, a new *Leo* command-line option, say `--write-black-sentinels`, will allow Leo to coexist with Black more easily.
